File: Python/Sets&Maps.py
from collections import defaultdict
import math
import copy
from collections import Counter

def accountSharing(connections) -> str:
    #Problem 30.1 - Account Sharing Detection
    
    usernameSet = set()
    
    for connection in connections:
        if connection[1] in usernameSet:
            return connection[0]
        
        usernameSet.add(connection[1])
    
    return ""

def mostSharedAccount(connections: list[tuple[str]]) -> str:
    #Problem 30.2 - Most Shared Account
    
    nameUsedCount = defaultdict(int)
    mostSharedUser = None
    maxNameCountUsed = 0
    
    for connection in connections:
        nameUsedCount[connection[1]] += 1
        if nameUsedCount[connection[1]] > maxNameCountUsed:
            mostSharedUser = connection[1]
            maxNameCountUsed = nameUsedCount[connection[1]]
    return mostSharedUser

def mostFrequentOctet(ips: list[str]) -> str:
    #Problem 30.3 - Most Frequent Octet
    
    octetFrequencyDictionary = defaultdict(int)
    mostCommonCountOctet = 0
    mostCommonOctet = None
    
    for ip in ips:
        octet = ip.partition('.')[0]
        octetFrequencyDictionary[octet] += 1
        if octetFrequencyDictionary[octet] > mostCommonCountOctet:
            mostCommonCountOctet = octetFrequencyDictionary[octet]
            mostCommonOctet = octet
        
    return mostCommonOctet

def multiAccountCheating(users: list[tuple]) -> bool:
    #Problem 30.4 - Multi-Account Cheating
    
    ipSet = set()
    
    for user in users:
        for ip in user[1]:
            if ip in ipSet:
                return True
        for ip in user[1]:
            ipSet.add(ip)
    
    return False

class DomainResolver:
    #Problem 30.5 - Domain Resolver
    
    def __init__(self):
        self.ipToDomainSetDictionary = defaultdict(set)
        self.domainToSubDomainDictionary = defaultdict(set)
    
    def registerDomain(self, ip: str, domain: str):
        self.ipToDomainSetDictionary[ip].add(domain)
    def registerSubdomain(self, domain: str, subdomain: str):
        self.domainToSubDomainDictionary[domain].add(subdomain)
    def hasSubdomain(self, ip: str, domain: str, subdomain: str) -> bool:
        if domain not in self.ipToDomainSetDictionary[ip]:
            return False
        if subdomain not in self.domainToSubDomainDictionary[domain]:
            return False
        
        return True

def findSquared(arr: list[int]) -> list[list[int]]:
    #Problem 30.6 - Find All Squares
    
    numberIndexDictionary = defaultdict(int)
    squaredIndex = []
    
    for index in range(len(arr)):
        numberIndexDictionary[arr[index] + 0.0] = index
        number = (arr[index] * arr[index]) + 0.0
        squareRoot = math.sqrt(arr[index])
        
        if number in numberIndexDictionary:
            squaredIndex.append([index, numberIndexDictionary[number]])
        elif squareRoot in numberIndexDictionary:
            squaredIndex.append([numberIndexDictionary[squareRoot], index])
    
    return squaredIndex

class Checker:
    #Problem 30.7 - Word Expansion Class
    
    def __init__(self, word1:str):
        self.word = word1        
        self.wordDictionaryFrequency = Counter(self.word)
        
    def expandsInto(self, word2: str) -> bool:
        if len(word2) <= len(self.word) or len(word2) > len(self.word) + 1:
            return False
        
        wordDictionaryFrequency1 = self.wordDictionaryFrequency.copy()
        
        wordLength = len(self.word)
        exactlyOneCount = 0
        
        for letter1 in word2:
            if letter1 not in wordDictionaryFrequency1:
                exactlyOneCount += 1
                if exactlyOneCount > 1:
                    return False
                continue
            
            wordDictionaryFrequency1[letter1] -= 1
    
            if wordDictionaryFrequency1[letter1] == -1:
                exactlyOneCount += 1
                if exactlyOneCount > 1:
                    return False
                continue
            
            wordLength -= 1
        
        if exactlyOneCount == 1 and wordLength == 0:
            return True
        
        return False

def suspectStudents(answers: list[str], m: int, students: list[tuple]) -> list[list[int]]:
    #Problem 30.8 - Cheater Detection
    
    if len(students) == 0 or m == 1 or len(answers) == 0:
        return []
    
    deskToTupleDictionary = defaultdict(tuple)
    deskToTupleDictionary[students[0][1]] = (students[0][0], students[0][2])
    
    suspectedStudentList = []
    
    for student in students[1:]:
        if student[1] - 1 in deskToTupleDictionary and (student[1] - 1) % m != 0:
            wrongAnswerIndexList = []
            for index in range(len(answers)):
                if answers[index] != student[2][index]:
                    wrongAnswerIndexList.append(index)
            
            for index1 in range(len(wrongAnswerIndexList)):
                if (student[2][wrongAnswerIndexList[index1]] == 
                    deskToTupleDictionary[student[1] - 1][1][wrongAnswerIndexList[index1]]):
                    suspectedStudentList.append([deskToTupleDictionary[student[1] - 1][0],student[0]])
                    break
                        
        if student[1] + 1 in deskToTupleDictionary and (student[1]  + 1) % m != 1:
            wrongAnswerIndexList = []
            for index in range(len(answers)):
                if answers[index] != student[2][index]:
                    wrongAnswerIndexList.append(index)
            
            for index1 in range(len(wrongAnswerIndexList)):
                if (student[2][wrongAnswerIndexList[index1]] == 
                    deskToTupleDictionary[student[1] + 1][1][wrongAnswerIndexList[index1]]):
                    suspectedStudentList.append([deskToTupleDictionary[student[1] + 1][0], student[0]])
                    break
            
        deskToTupleDictionary[student[1]] = (student[0], student[2])
    
    return suspectedStudentList

def alphabeticSumProduct(words: list[str], target: int) -> bool:
    #Problem 30.9 - Product of Alphabetical Sums
    
    if len(words) == 0:
        return False
    
    targetMultiplesSet = set()
    
    for word in words:
        total = 0
        for letter in word:
            total += ord(letter) - 96
            
        if target % total == 0:
            targetMultiplesSet.add(total)
    
    for number0 in targetMultiplesSet:
        for number1 in targetMultiplesSet:
            if target / (number0 * number1) in targetMultiplesSet:
                return True
    
    return False

#TESTS

def runAccountSharingDetectionTests():
    tests = [
        # Example 1 
      	([("203.0.113.10", "mike"), ("298.51.100.25", "bob"),
        ("292.0.2.5", "mike"), ("203.0.113.15", "bob2")], "203.0.113.10"),
      	# Example 2 
      	([("111.0.0.0", "mike"), ("111.0.0.1", "mike"),
        ("111.0.0.2", "bob"), ("111.0.0.3", "bob")], "111.0.0.0"),
      	# Example 3 
      	([("111.0.0.0", "mike"), ("111.0.0.1", "mike2"),
        ("111.0.0.2", "mike3"), ("111.0.0.3", "mike4")], ""),
      	# Edge case - empty list
      	([], ""),
      	# Edge case - single connection
      	([("1.1.1.1", "alice")], ""),
  	]
    
    for connections, want in tests:
        got = accountSharing(connections)
        
        # Check if got matches want directly
        if got == want:
         continue
     
     	# If want is empty, got must also be empty
        if want == "":
            assert got == "", f"\naccount_sharing({connections}): got: {got}, want: {want}\n"
            continue
	
    print("ALL ACCOUNT SHARING DETECTION TESTS PROVIDED PASSED.")

def runMostSharedAccountTests():
    tests = [
        # Example 
        ([("203.0.113.10", "mike"), ("208.51.100.25", "bob"),
        ("202.0.2.5", "mike"), ("203.0.113.15", "bob2")], "mike"),
        # Additional test cases
        ([], None),
        ([("1.1.1.1", "alice")], "alice"),
        ([("1.1.1.1", "alice"), ("1.1.1.2", "bob"),
        ("1.1.1.3", "alice"), ("1.1.1.4", "bob")], "alice"),
    ]
    
    for connections, want in tests:
        got = mostSharedAccount(connections)
        assert got == want or (want and got and
                           len([(ip, u) for ip, u in connections if u == got]) ==
                           len([(ip, u) for ip, u in connections if u == want])), \
            f"\nmost_shared_account({connections}): got: {got}, want: {want}\n"
    
    print("ALL MOST SHARED ACCOUNT TESTS PROVIDED PASSED.")

def runMostFrequentOctetTests():
    tests = [
        # Example 
        (["203.0.113.10", "208.51.100.5", "202.0.2.5", "203.0.113.5"], "203"),
        # Additional test cases
        ([], None),
        (["192.168.1.1"], "192"),
        (["10.0.0.1", "10.0.0.2", "192.168.1.1"], "10"),
        (["172.16.0.1", "172.16.0.2", "172.17.0.1", "172.16.0.3"], "172"),
    ]
    
    for ips, want in tests:
        got = mostFrequentOctet(ips)
        assert got == want, f"\nmost_frequent_octet({ips}): got: {got}, want: {want}\n"
        
    print("ALL MOST FREQUENT OCTET TESTS PROVIDED PASSED.")

def runMultiAccountCheatingTests():
    tests = [
        # Example 
        ([("mike", ["203.0.3.10", "208.51.0.5", "52.0.2.5"]),
        ("bob", ["111.0.0.10", "222.0.0.5", "222.0.0.8"]),
          ("bob2", ["222.0.0.5", "222.0.0.8", "111.0.0.10"])], True),
        # Additional test cases
        ([], False),
        ([("alice", ["1.1.1.1"])], False),
        ([("alice", ["1.1.1.1", "2.2.2.2"]),
        ("bob", ["2.2.2.2", "1.1.1.1"])], True),
        ([("alice", ["1.1.1.1"]), ("bob", ["2.2.2.2"])], False),
    ]
    
    for users, want in tests:
        got = multiAccountCheating(users)
        assert got == want, f"\nmulti_account_cheating({users}): got: {got}, want: {want}\n"
    
    print("ALL MULTI-ACCOUNT CHEATING TESTS PROVIDED PASSED.")

def runDomainResolverTests():
    tests = [
        # Example 
        (
            [
                ("register_domain", "192.168.1.1", "example.com"),
                ("register_domain", "192.168.1.1", "example.org"),
                ("register_domain", "192.168.1.2", "domain.com"),
                ("register_subdomain", "example.com", "a"),
                ("register_subdomain", "example.com", "b"),
                ("has_subdomain", "192.168.1.1", "example.com", "a"),
                ("has_subdomain", "192.168.1.1", "example.com", "c"),
                ("has_subdomain", "127.0.0.1", "example.com", "a"),
                ("has_subdomain", "192.168.1.1", "example.org", "a"),
                ("has_subdomain", "192.168.1.2", "example.com", "a"),
            ],
            [None, None, None, None, None, True, False, False, False, False]
        ),
        # Additional test cases
        (
            [
                ("register_domain", "1.1.1.1", "test.com"),
                ("register_subdomain", "test.com", "www"),
                ("has_subdomain", "1.1.1.1", "test.com", "www"),
            ],
            [None, None, True]
        ),
        (
            [
                ("register_domain", "1.1.1.1", "site1.com"),
                ("register_domain", "2.2.2.2", "site2.com"),
                ("register_subdomain", "site1.com", "www"),
                ("register_subdomain", "site2.com", "www"),
                ("has_subdomain", "1.1.1.1", "site1.com", "www"),  # Should be True
                ("has_subdomain", "2.2.2.2", "site2.com", "www"),  # Should be True
                ("has_subdomain", "1.1.1.1", "site2.com", "www"),  # Should be False (wrong IP)
                ("has_subdomain", "2.2.2.2", "site1.com", "www"),  # Should be False (wrong IP)
            ],
            [None, None, None, None, True, True, False, False]
        ),
    ]
    
    for operations, wants in tests:
        resolver = DomainResolver()
        for i, op in enumerate(operations):
            if op[0] == "register_domain":
                got = resolver.registerDomain(op[1], op[2])
            elif op[0] == "register_subdomain":
                got = resolver.registerSubdomain(op[1], op[2])
            else:  # has_subdomain
                got = resolver.hasSubdomain(op[1], op[2], op[3])
            want = wants[i]
            assert got == want, \
                f"\n{op[0]}({', '.join(repr(x) for x in op[1:])}): got: {got}, want: {want}\n"
    
    print("ALL DOMAIN RESOLVER TESTS PROVIDED PASSED.")

def runFindAllSquaresTests():
    tests = [
        # Example 
        ([4, 10, 3, 100, 5, 2, 10000], [[5, 0], [1, 3], [3, 6]]),
        # Additional test cases
        ([], []),
        ([1], [[0, 0]]),
        ([2, 4], [[0, 1]]),
    ]
    
    for arr, want in tests:
        got = findSquared(arr)
        # Sort both lists to compare them regardless of order
        got.sort()
        want.sort()
        assert got == want, f"\nfind_squared({arr}): got: {got}, want: {want}\n"

    print("ALL FIND ALL SQUARES TESTS PROVIDED PASSED.")

def runWordExpansionClassTests():
    tests = [
        # Example 1 
        (
            ("tea", [
                ("tea", False),
                ("team", True),
                ("seam", False),
                ]
             )
        ),        
        # Example 2 
        (
            ("on", [
                ("nooo", False),
                ("not", True),
                ("now", True),
                ]
            )
        ),
        # Additional test cases
        (
            ("", [
                ("a", True),
                ("", False),
                ("ab", False),
                ]
            )
        ),
        
        (
            ("xyz", [
                ("wxyz", True),
                ("xyzw", True),
                ("xyza", True),
                ("xyz", False),
                ]
             )
        ),
    ]
    
    for s, checks in tests:
        checker = Checker(s)
        for s2, want in checks:
            got = checker.expandsInto(s2)
            assert got == want, \
                f"\nChecker({repr(s)}).expandsInto({repr(s2)}): got: {got}, want: {want}\n"
    
    print("ALL WORD EXPANSION CLASS TESTS PROVIDED PASSED.")
    
def runCheaterDetectionTests():
    tests = [
        # Example 
        (
            ['a', 'b', 'c', 'c'],
            5,
            [
                (4, 10, ['a', 'b', 'c', 'd']),
                (1, 6, ['a', 'b', 'c', 'd']),
                (3, 8, ['a', 'b', 'd', 'd']),
                (5, 11, ['a', 'b', 'c', 'd']),
                (9, 7, ['a', 'b', 'c', 'd']),
                (6, 16, ['a', 'b', 'd', 'd']),
            ],
            [[1, 9], [3, 9]]
        ),
        # Additional test cases
        (
            ['a', 'b'],
            2,
            [
                (1, 1, ['a', 'b']),
                (2, 2, ['a', 'b']),
            ],
            []  # Perfect scores are not suspicious
        ),
        (
            ['a', 'b'],
            2,
            [
                (1, 1, ['b', 'b']),
                (2, 2, ['b', 'b']),
            ],
            [[1, 2]]
        ),
        (
            ['a', 'b'],
            2,
            [
                (1, 1, ['b', 'b']),
                (2, 3, ['b', 'b']),
            ],
            []  # Different rows
        ),
    ]

    for answers, m, students, want in tests:
        got = suspectStudents(answers, m, students)
        # Sort both lists to compare them regardless of order
        assert got == want, \
            f"\nsuspect_students({answers}, {m}, {students}): got: {got}, want: {want}\n"
            
    print("ALL CHEATER DETECTION TESTS PROVIDED PASSED.")

def runProductOfAlphabeticalSumsTests():
    tests = [
        # Example 1 
        (["abc", "fg", "hij", "klm", "nop", "qrs", "vwx"], 1620, True),
        # Example 2 
        (["a", "b"], 2, True),
        # Additional test cases
        ([], 1, False),
        (["a"], 1, True),
        (["a", "b", "c"], 6, True),
        (["a", "b", "c"], 7, False),
    ]
    
    for words, target, want in tests:
        got = alphabeticSumProduct(words, target)
        assert got == want, \
            f"\nalphabetic_sum_product({words}, {target}): got: {got}, want: {want}\n"
    
    print("ALL PRODUCT OF ALPHABETICAL SUMS TESTS PROVIDED PASSED.")
   
#ALL TESTS
    
def RunAllSetsAndMapsTests():
    runAccountSharingDetectionTests()
    runMostSharedAccountTests()
    runMostFrequentOctetTests()
    runMultiAccountCheatingTests()
    runFindAllSquaresTests()
    runWordExpansionClassTests()
    runCheaterDetectionTests()
    runProductOfAlphabeticalSumsTests()
    
    print()
    print("------------------------------------------------------")
    print("ALL INCLUDED SETS AND MAPS TESTS IN THE FILE PASSED. |")
    print("------------------------------------------------------")

if __name__ == "__main__":
    RunAllSetsAndMapsTests()

File: C++/C++/Sets&Maps.cpp
//
//  Sets&Maps.cpp
//  C++
//
//  Created by Richmond Laureta on 6/8/26.
//

#include "Header.hpp"

string accountSharing(const vector<pair<string, string>> &connections)
{
    //Problem 30.1 - Account Sharing Detection
    
    unordered_set<string> names;
    
    for(pair<string, string> connection: connections)
    {
        if(names.find(connection.second) != names.end()) return connection.first;
        
        names.insert(connection.second);
    }
    
    return "";
}


string mostSharedAccount(const vector<pair<string, string>> &connections)
{
    //Problem 30.2 - Most Shared Account
    
    unordered_map<string, int> nameUsedCount;
    string mostSharedUser = "";
    int maxNameCountUsed = 0;
    
    for(pair<string, string> connection: connections)
    {
        nameUsedCount[connection.second]++;
        if(nameUsedCount[connection.second] > maxNameCountUsed)
        {
            mostSharedUser = connection.second;
            maxNameCountUsed = nameUsedCount[connection.second];
        }
    }
    
    return mostSharedUser;
}

string mostFrequentOctet(const vector<string> &ips)
{
    //Problem 30.3 - Most Frequent Octet
    
    unordered_map<string, int> octetFrequencyMap;
    int mostCommonCountOctet = 0;
    string mostCommonOctet = "";
    
    for(string ip: ips)
    {
        size_t dotIndex = ip.find('.');
        string octet = ip.substr(0, dotIndex);
        octetFrequencyMap[octet]++;
        if(octetFrequencyMap[octet] > mostCommonCountOctet)
        {
            mostCommonCountOctet = octetFrequencyMap[octet];
            mostCommonOctet = octet;
        }
    }
    
    return mostCommonOctet;
}

bool multiAccountCheating(const vector<pair<string, vector<string>>> &users)
{
    //Problem 30.4 - Multi-Account Cheating
    
    unordered_set<string> ipSet = {};
    
    for(pair<string, vector<string>> user: users)
    {
        for(string ip: user.second)
        {
            if(ipSet.find(ip) != ipSet.end()) return true;
        }
        
        for(string ip: user.second)
        {
            ipSet.insert(ip);
        }
    }
    
    return false;
}

void DomainResolver::registerDomain(string ip, string domain)
{
    //Problem 30.5 - Domain Resolver
    
    ipToDomainSetMap[ip].insert(domain);
}

void DomainResolver::registerSubdomain(string domain, string subdomain)
{
    //Problem 30.5 - Domain Resolver
    
    domainToSubdomainSetMap[domain].insert(subdomain);
}

bool DomainResolver::hasSubdomain(string ip, string domain, string subdomain)
{
    //Problem 30.5 - Domain Resolver
    
    if(ipToDomainSetMap[ip].find(domain) == ipToDomainSetMap[ip].end()) return false;
    
    if(domainToSubdomainSetMap[domain].find(subdomain) == domainToSubdomainSetMap[domain].end()) return false;
        
    return true;
}

vector<vector<int>> findSquared(const vector<int> &arr)
{
    //Problem 30.6 - Find All Squares
    
    unordered_map<double, int> numberIndexMap;
    vector<vector<int>> squaredIndex;
    
    for(int index = 0; index < (int) arr.size(); index++)
    {
        numberIndexMap[arr[index] + 0.0] = index;

        double number = (arr[index] * arr[index]) + 0.0;
        double squareRoot = sqrt(arr[index]);
        
        if(numberIndexMap.contains(number)) squaredIndex.push_back({index, numberIndexMap[number]});
        else if(numberIndexMap.contains(squareRoot)) squaredIndex.push_back({numberIndexMap[squareRoot], index});
        
    }
    
    return squaredIndex;
}

Checker::Checker(const string &word1)
{
    //Problem 30.7 - Word Expansion Class
    
    this->word = word1;
    
    for(char letter: word1)
    {
        wordMapFrequency[letter]++;
    }
}

bool Checker::expandsInto(const string &word2)
{
    //Problem 30.7 - Word Expansion Class
    
    if((int) word2.length() <= (int) this->word.length() or (int) word2.length() > ((int) word.length()) + 1) return false;
    
    unordered_map<char, int> wordFrequencyMap1;
    wordFrequencyMap1.reserve(this->wordMapFrequency.size());
    wordFrequencyMap1 = this->wordMapFrequency;
    
    int wordLength = (int) word.length();
    int exactlyOneCount = 0;
    
    for(char letter1: word2)
    {
        if(wordFrequencyMap1.find(letter1) == wordFrequencyMap1.end())
        {
            exactlyOneCount++;
            if(exactlyOneCount > 1) return false;
            continue;
        }
        
        wordFrequencyMap1[letter1]--;
        
        if(wordFrequencyMap1[letter1] == -1)
        {
            exactlyOneCount++;
            if(exactlyOneCount > 1) return false;
            continue;
        }
        
        wordLength--;
    }
    
    if(exactlyOneCount == 1 and wordLength == 0) return true;
    
    return false;
}

vector<vector<int>> suspectStudents(const vector<char> &answers, int m, const vector<Student> &students)
{
    //Problem 30.8 - Cheater Detection
    
    if((int) students.size() == 0 or m == 1 or (int) answers.size() == 0) return {};

    unordered_map<int, tuple<int, vector<char>>> deskToTupleMap;
    
    vector<vector<int>> suspectedPairsVector;
    
    for(Student student: students)
    {
        if(deskToTupleMap.find(student.desk - 1) != deskToTupleMap.end() and (student.desk - 1) % m != 0)
        {
            vector<int> wrongAnswerIndexVector = {};
            
            for(int index = 0; index < (int) answers.size(); index++)
            {
                if(answers[index] != student.answers[index]) wrongAnswerIndexVector.push_back(index);
            }
            
            for(int index1 = 0; index1 < (int) wrongAnswerIndexVector.size(); index1++)
            {
                if(student.answers[wrongAnswerIndexVector[index1]] ==
                   get<1>(deskToTupleMap[student.desk - 1])[wrongAnswerIndexVector[index1]])
                {
                    suspectedPairsVector.push_back({get<0>(deskToTupleMap[student.desk - 1]), student.ID});
                    break;
                }
            }
        }
        
        if(deskToTupleMap.find(student.desk + 1) != deskToTupleMap.end() and (student.desk + 1) % m != 1)
        {
            vector<int> wrongAnswerIndexVector = {};
            
            for(int index = 0; index < (int) answers.size(); index++)
            {
                if(answers[index] != student.answers[index]) wrongAnswerIndexVector.push_back(index);
            }
            
            for(int index1 = 0; index1 < (int) wrongAnswerIndexVector.size(); index1++)
            {
                if(student.answers[wrongAnswerIndexVector[index1]] ==
                   get<1>(deskToTupleMap[student.desk + 1])[wrongAnswerIndexVector[index1]])
                {
                    suspectedPairsVector.push_back({get<0>(deskToTupleMap[student.desk + 1]), student.ID});
                    break;
                }
            }
        }
        
        deskToTupleMap[student.desk] = {student.ID, student.answers};
    }
    
    return suspectedPairsVector;
}

bool alphabeticSumProduct(const vector<string> &words, int target)
{
    //Problem 30.9 - Product of Alphabetical Sums
    
    if((int) words.size() == 0) return false;
    
    unordered_set<double> targetMultiplesSet = {};
    
    for(string word: words)
    {
        int total = 0;
        for(char letter: word) total += static_cast<int>(letter) - 96 + 0.0;
        
        if(target % total == 0) targetMultiplesSet.insert(total);
    }
    
    for(int number0: targetMultiplesSet)
    {
        for(int number1: targetMultiplesSet)
        {
            if(targetMultiplesSet.find(target / (number0 * number1)) != targetMultiplesSet.end()) return true;
        }
    }
    
    return false;
}

Action::Action(string a, string act, int t) : agent(a), action(act), ticket(t)
{
    //Problem 30.10 - Action Log Anomalies
    
};

vector<int> findAnomalies(const vector<Action> &log)
{
    //Problem 30.10 - Action Log Anomalies
    
    unordered_set<int> anomalyTicketSet = {};
    unordered_set<int> stillOpentTicketSet = {};
    unordered_set<int> openedAndClosedTicketSet = {};
    unordered_set<string> agentsNeedToCloseSet = {};
    
    unordered_map<int, string> ticketToAgentMap;
    unordered_map<string, int> agentToTicketMap;
    
    vector<int> anomalyTicketVector = {};
    
    for(Action entry: log)
    {
        if(anomalyTicketSet.find(entry.ticket) != anomalyTicketSet.end())
        {
            if(agentsNeedToCloseSet.find(entry.agent) != agentsNeedToCloseSet.end() and
               entry.ticket != agentToTicketMap[entry.agent] and
               anomalyTicketSet.find(agentToTicketMap[entry.agent]) == anomalyTicketSet.end())
            {
                anomalyTicketSet.insert(agentToTicketMap[entry.agent]);
                anomalyTicketVector.push_back(agentToTicketMap[entry.agent]);
            }
            
            if(entry.action == "open")
            {
                agentsNeedToCloseSet.insert(entry.agent);
                stillOpentTicketSet.insert(entry.ticket);
                agentToTicketMap[entry.agent] = entry.ticket;
                ticketToAgentMap[entry.ticket] = entry.agent;
            }
            
            continue;
        }
        
        if(openedAndClosedTicketSet.find(entry.ticket) != openedAndClosedTicketSet.end() and
           anomalyTicketSet.find(entry.ticket) == anomalyTicketSet.end())
        {
            anomalyTicketSet.insert(entry.ticket);
            anomalyTicketVector.push_back(entry.ticket);
            continue;
        }
        
        if(ticketToAgentMap[entry.ticket] == "" and entry.action == "close")
        {
            if(agentsNeedToCloseSet.find(entry.agent) != agentsNeedToCloseSet.end())
            {
                anomalyTicketSet.insert(agentToTicketMap[entry.agent]);
                anomalyTicketVector.push_back(agentToTicketMap[entry.agent]);
                stillOpentTicketSet.erase(agentToTicketMap[entry.agent]);
            }
            
            anomalyTicketVector.push_back(entry.ticket);
            anomalyTicketSet.insert(entry.ticket);
            continue;
        }
        else if(ticketToAgentMap[entry.ticket] == "" and entry.action == "open")
        {
            if(agentsNeedToCloseSet.find(entry.agent) != agentsNeedToCloseSet.end())
            {
                anomalyTicketSet.insert(agentToTicketMap[entry.agent]);
                anomalyTicketVector.push_back(agentToTicketMap[entry.agent]);
                stillOpentTicketSet.erase(agentToTicketMap[entry.agent]);
            }
            
            ticketToAgentMap[entry.ticket] = entry.action;
            stillOpentTicketSet.insert(entry.ticket);
            agentToTicketMap[entry.agent] = entry.ticket;
            agentsNeedToCloseSet.insert(entry.agent);
            continue;
        }
        
        if(ticketToAgentMap[entry.ticket] != "" and entry.action == "close")
        {
            if(entry.agent != ticketToAgentMap[entry.ticket])
            {
                anomalyTicketSet.insert(entry.ticket);
                anomalyTicketVector.push_back(entry.ticket);
                agentsNeedToCloseSet.erase(ticketToAgentMap[entry.ticket]);
                stillOpentTicketSet.erase(entry.ticket);
                continue;
            }
            
            stillOpentTicketSet.erase(entry.ticket);
            openedAndClosedTicketSet.insert(entry.ticket);
            agentsNeedToCloseSet.erase(entry.agent);
        }
    }
    
    for(int ticket: stillOpentTicketSet)
    {
        if(anomalyTicketSet.find(ticket) == anomalyTicketSet.end()) anomalyTicketVector.push_back(ticket);
    }
    
    return anomalyTicketVector;
}

std::unordered_set<int> setIntersection(
    const std::unordered_set<int>& set1,
    const std::unordered_set<int>& set2) {
  std::unordered_set<int> result;
  // Iterate through the smaller set for efficiency
  if (set1.size() <= set2.size()) {
    for (int x : set1) {
      if (set2.find(x) != set2.end()) {
        result.insert(x);
      }
    }
  } else {
    for (int x : set2) {
      if (set1.find(x) != set1.end()) {
        result.insert(x);
      }
    }
  }
  return result;
}

int largestSetIntersectionPrefixSum(const std::vector<std::vector<int>>& sets) {
  int n = (int) sets.size();
  if (n == 1) {
    return 0;
  }

  std::vector<std::unordered_set<int>> hashSets(n);
  for (int i = 0; i < n; i++) {
    hashSets[i] = std::unordered_set<int>(sets[i].begin(), sets[i].end());
  }

  // Compute prefix intersections
  std::vector<std::unordered_set<int>> prefixIntersections(n);
  prefixIntersections[0] = hashSets[0];
  for (int i = 1; i < n; i++) {
    prefixIntersections[i] = setIntersection(prefixIntersections[i - 1], hashSets[i]);
  }

  // Compute suffix intersections
  std::vector<std::unordered_set<int>> suffixIntersections(n);
  suffixIntersections[n - 1] = hashSets[n - 1];
  for (int i = n - 2; i >= 0; i--) {
    suffixIntersections[i] = setIntersection(suffixIntersections[i + 1], hashSets[i]);
  }

  // Find the best index to exclude
  int bestIndex = 0;
  int maxSize = 0;

  for (int i = 0; i < n; i++) {
    // Compute intersection excluding sets[i]
    std::unordered_set<int> intersection;
    if (i == 0) {
      intersection = suffixIntersections[1];
    } else if (i == n - 1) {
      intersection = prefixIntersections[n - 2];
    } else {
      intersection = setIntersection(prefixIntersections[i - 1], suffixIntersections[i + 1]);
    }

    if (intersection.size() > maxSize) {
      maxSize = (int) intersection.size();
      bestIndex = i;
    }
  }

  return bestIndex;
}

int largestSetIntersectionFrequencyMap(const vector<vector<int>> &sets)
{
    //Problem 30.11 - Largest Set Intersection
    
    if((int) sets.size() == 1) return 0;
    
    unordered_map<int, int> numberToFrequencyMap;
    
    for(vector<int> set: sets)
    {
        for(int number: set)
        {
            numberToFrequencyMap[number]++;
        }
    }
    
    int index = 0;
    int minimumIntersectionIndex = 0;
    int minimumIntersectionCount = INT_MAX;
    
    for(vector<int> set: sets)
    {
        int totalIntersection = 0;
        for(int number: set)
        {
            totalIntersection += numberToFrequencyMap[number] - 1;
        }
        
        if(totalIntersection == minimumIntersectionCount and ((int) sets[index].size()) < ((int) sets[minimumIntersectionIndex].size())) minimumIntersectionIndex = index;
        else if(totalIntersection < minimumIntersectionCount)
        {
            minimumIntersectionIndex = index;
            minimumIntersectionCount = totalIntersection;
        }
        
        index++;
    }
    
    return minimumIntersectionIndex;
}
